### PR TITLE
[isoltest] Enforcing compileViaYul to be set if test can pass via yul

### LINKED ISSUE #8686: isoltest: throw failure if input can be compiled via yul (--ir) but was marked as `compileViaYul: false`
Idea from https://github.com/ethereum/solidity/pull/8554#issuecomment-614326427

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #8686

### CODE REVIEW

File: test/tools/isoltest.cpp [hunk 1]
```diff
@@ -189,12 +189,24 @@ TestTool::Result TestTool::process()
 	}
 	catch (boost::exception const& _e)
 	{
+		auto const* semanticTest = dynamic_cast<SemanticTest const*>(m_test.get());
+		if (semanticTest && semanticTest->runningEnforcedViaYul())
+		{
+			AnsiColorized(cout, formatted, {BOLD, GREEN}) << "OK" << endl;
+			return Result::Success;
+		}
```
Comment (mijovic) [inline test/tools/isoltest.cpp]: I didn't find better solution for this. If there is missing feature in sol2yul it usually asserts which ends up here.
When doing forced run of compilation via yul it sometimes triggers exception...

@axic if you have any better idea, it is very welcome

Comment (mijovic) [reply test/tools/isoltest.cpp]: Also, enforced run is last one in the chain for particular test, so it is fine to return success in this case

**Comment by Marenz** [comment]
One test is failing

**Comment by Marenz** [comment]
I wonder if the failure is caused by running more tests and it's showing us a problem in the yul code

**Comment by mijovic** [comment]
> One test is failing

Tests are fine, all are passing. It has some memory leak for generate function in IRGenerator, will check it now. Not sure how it is detected only in t_ubu_asan_constantinople

File: test/libsolidity/SemanticTest.cpp [hunk 2]
```diff
@@ -78,8 +78,10 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 
 TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePrefix, bool _formatted)
 {
-	for(bool compileViaYul: set<bool>{!m_runWithoutYul, m_runWithYul})
+
+	for(bool compileViaYul: set<bool>{!m_runWithoutYul, m_runWithYul, true})
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]:
```suggestion
	for (bool compileViaYul: set<bool>{!m_runWithoutYul, m_runWithYul, true})
```

File: test/boostTest.cpp [hunk 3]
```diff
@@ -109,7 +110,9 @@ int registerTests(
 					}
 					catch (boost::exception const& _e)
 					{
-						BOOST_ERROR("Exception during extracted test: " << boost::diagnostic_information(_e));
+						auto const* semanticTest = dynamic_cast<SemanticTest const*>(testCase.get());
```
Comment (chriseth) [inline test/boostTest.cpp]: I think we should instead have a try/catch inside the semantic test.

Comment (mijovic) [reply test/boostTest.cpp]: I agree, was thinking to do like that

File: test/libsolidity/SemanticTest.cpp [hunk 4]
```diff
@@ -155,7 +157,14 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
 			}
 		}
 
-		if (!success)
+		if (success && !m_runWithYul && compileViaYul)
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]: Is it possible to make this just a regular expectation failure and thus pressing `u(update)` auto-updates the setting?

**Comment by mijovic** [comment]
Tests are showing that some of sol2yul features are leaking memory, will try to fix that as part of this PR

File: test/tools/isoltest.cpp [hunk 5]
```diff
@@ -189,12 +189,24 @@ TestTool::Result TestTool::process()
...
 		AnsiColorized(cout, formatted, {BOLD, RED}) <<
 			"Exception during test: " << boost::diagnostic_information(_e) << endl;
 		return Result::Exception;
 	}
 	catch (std::exception const& _e)
 	{
+		auto const* semanticTest = dynamic_cast<SemanticTest const*>(m_test.get());
```
Comment (chriseth) [inline test/tools/isoltest.cpp]: Same here - this should happen inside SemanticTest.

File: test/libsolidity/SemanticTest.cpp (hunk 4 above)
Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp]: I will check how to do this, it would be great to have auto update on this feature

Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp]: Figured it out, done with it

File: test/libsolidity/SemanticTest.cpp [hunk 6]
```diff
@@ -78,112 +78,141 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
...
 			}
-			else if (!compileViaYul && m_runWithYul)
-				AnsiColorized(_stream, _formatted, {BOLD, YELLOW}) << _linePrefix << endl << _linePrefix
-					<< "Note that the test also has to pass via Yul." << endl;
-			return TestResult::Failure;
+		}
+		catch (boost::exception const& _e)
+		{
+			if (compileViaYul && !m_runWithYul)
+				return TestResult::Success;
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp]: Why return? Shouldn't this just be a `break`?

File: test/libsolidity/SemanticTest.cpp [hunk 7]
```diff
@@ -204,6 +233,20 @@ void SemanticTest::printUpdatedExpectations(ostream& _stream, string const&) con
...
 }
 
+void SemanticTest::printUpdatedSettings(ostream& _stream, string const& _linePrefix)
+{
+	auto& settings = m_reader.settings();
+	if (settings.empty() && !m_compileViaYulCanBeSet)
+		return;
+
+	_stream << _linePrefix << "// ====" << endl;
+	if (m_compileViaYulCanBeSet)
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp:248]: Is this ever true if we run `compileViaYul: only`?

File: test/libsolidity/SemanticTest.cpp (hunk 6 above)
Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp]: Indeed, I'll change to break

File: test/libsolidity/SemanticTest.cpp (hunk 7 above)
Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp:248]: It is never true in that case. It is true only if it is missing compileViaYul, or if it is explicitly set to false. Anyway if it makes confusion, I can try making it more readable

File: test/libsolidity/SemanticTest.cpp (hunk 6 above)
Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp]: I did continue

File: test/libsolidity/SemanticTest.cpp (hunk 7 above)
Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp:248]: it is set to false when we start executing test and it is set to true with
```
if (success && !m_runWithYul && compileViaYul)
{
	m_compileViaYulCanBeSet = true;
	...
}
```

So, it will be set only if test succeded with compileViaYul and m_runWithYul wasn't selected, so should be fine

File: test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol [hunk 8]
```diff
@@ -7,10 +7,9 @@ contract C {
         return x(a);
     }
 
-    function g(uint256 x) public returns (uint256) {
+    function g(uint256 x) public pure returns (uint256) {
```
Comment (axic) [inline test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol:10]: Why is this change here now?

File: test/libsolidity/SemanticTest.cpp [hunk 9]
```diff
@@ -78,112 +78,141 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
...
+					AnsiColorized(_stream, _formatted, {BOLD, YELLOW}) << _linePrefix << endl << _linePrefix
+						<< "Note that the test also has to pass via Yul." << endl;
+				return TestResult::Failure;
 			}
-			else if (!compileViaYul && m_runWithYul)
-				AnsiColorized(_stream, _formatted, {BOLD, YELLOW}) << _linePrefix << endl << _linePrefix
-					<< "Note that the test also has to pass via Yul." << endl;
-			return TestResult::Failure;
+		}
+		catch (boost::exception const& _e)
```
Comment (Marenz) [inline test/libsolidity/SemanticTest.cpp]: the `_e` parameter is unused and causes an error

File: test/libsolidity/SemanticTest.cpp [hunk 10]
```diff
@@ -78,112 +78,141 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
...
+		{
+			if (compileViaYul && !m_runWithYul)
+				continue;
+
+			throw;
+		}
+		catch (std::exception const& _e)
+		{
+			if (compileViaYul && !m_runWithYul)
+				continue;
```
Comment (Marenz) [inline test/libsolidity/SemanticTest.cpp:213]: the `_e` parameter is unused and causes an error

Comment (mijovic) [reply test/libsolidity/SemanticTest.cpp:213]: I removed it now, will rebase, check if additional tests are activated and push

File: test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol (hunk 8 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol:10]: Something weird happens, without pure, it passes with compileViaYul in isoltest, but fails in soltest with --optimize (it triggers warning and somehow fails later on)

with pure, it doesn't pass viaYul even in isoltest

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 11]
```diff
@@ -266,22 +266,26 @@ pair<string, map<ContractDefinition const*, string>> IRGenerator::evaluateConstr
...
-				);
+			))
+				if (
+					FunctionDefinition const* baseConstructor = baseContract->constructor();
+					baseConstructor && modifier->arguments()
+				)
+					baseConstructorArguments.emplace_back(
+						dynamic_cast<ContractDefinition const*>(baseConstructor->scope()),
+						modifier->arguments()
+					);
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Maybe should be extracted as a fix for constructor PR previously merged

Comment (axic) [reply libsolidity/codegen/ir/IRGenerator.cpp]: You mean #8591 ?

File: libyul/optimiser/ASTCopier.cpp [hunk 12]
```diff
@@ -47,10 +47,12 @@ Statement ASTCopier::operator()(VariableDeclaration const& _varDecl)
 
 Statement ASTCopier::operator()(Assignment const& _assignment)
 {
+	// throws and causes memleak in translateVector call, changed order to throw before allocation
```
Comment (axic) [inline libyul/optimiser/ASTCopier.cpp]: Add FIXME, TODO or @todo here?

Comment (Marenz) [reply libyul/optimiser/ASTCopier.cpp]: this is already fixed by changing the order as the comment says

Comment (Marenz) [reply libyul/optimiser/ASTCopier.cpp]: comment is maybe a bit ambiguous

File: test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol (hunk 8 above)
Comment (axic) [reply test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol:10]: Can you isolate this case and open a tracking issue?

File: libyul/optimiser/ASTCopier.cpp (hunk 12 above)
Comment (axic) [reply libyul/optimiser/ASTCopier.cpp]: Yes but in cases where such order is required to maintain working code it has to be highlighted.

Comment (Marenz) [reply libyul/optimiser/ASTCopier.cpp]: You mean more highlighted than with the given comment?

Comment (axic) [reply libyul/optimiser/ASTCopier.cpp]: The comment is not very clear to me tbh.

**Comment by leonardoalt** [comment]
asan mem leak

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 11 above)
Comment (mijovic) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Yes, this part here is a fix for that PR

File: test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol (hunk 8 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol:10]: I can open issue, but I do not fully understand what does it mean to isolate test case?

**Comment by axic** [comment]
And
```
Slowest test: ABIEncoderTest bytesNN_arrays_dyn (took 179.38 seconds).
```

**Comment by mijovic** [comment]
There are few more places where same leak happens, looking to solve them as well, and will do another update to PR

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 11 above)
Comment (mijovic) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Moved to another PR, will leave it here as well until other PR is merged. Is needed for CI to successfully completes all jobs

**Comment by chriseth** [comment]
Oh nice! It worked! Unfortunately, I merged the other PR first....

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
🎉

File: test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol (hunk 8 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/functionTypes/pass_function_types_internally.sol:10]: For the record: the change here causes IR codegen to generate invalid Yul, that's why enforce-via-yul starts to ignore this. Without this change, there was a Yul parser error that bubbled up as weird ICE error - but that should be provisionally fixed by https://github.com/ethereum/solidity/pull/8845 now.
However, we should investigate the invalid yul code generated, when the version *with* ``pure`` is compiled via yul.
